File: src-tauri/src/commands/ai/copilot.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Deserialize)]
pub struct CopilotStartLoginRequest {
    pub client_id: String,
}

#[derive(Debug, Serialize)]
pub struct CopilotStartLoginResponse {
    pub device_code: String,
    pub user_code: String,
    pub verification_uri: String,
    pub expires_in: u64,
    pub interval: u64,
}

#[tauri::command]
pub async fn copilot_start_device_login(
    req: CopilotStartLoginRequest,
) -> Result<CopilotStartLoginResponse, String> {
    if req.client_id.trim().is_empty() {
        return Err("GitHub OAuth client ID is required".to_string());
    }

    let result = crate::ai::auth::start_device_flow(&req.client_id)
        .await
        .map_err(|e| e.to_string())?;

    crate::ai::auth::store_client_id(&req.client_id).map_err(|e| e.to_string())?;

    Ok(CopilotStartLoginResponse {
        device_code: result.device_code,
        user_code: result.user_code,
        verification_uri: result.verification_uri,
        expires_in: result.expires_in,
        interval: result.interval,
    })
}

#[derive(Debug, Deserialize)]
pub struct CopilotPollLoginRequest {
    pub client_id: String,
    pub device_code: String,
}

#[derive(Debug, Serialize)]
pub struct CopilotPollLoginResponse {
    pub authorized: bool,
}

#[tauri::command]
pub async fn copilot_poll_device_login(
    req: CopilotPollLoginRequest,
) -> Result<CopilotPollLoginResponse, String> {
    if req.client_id.trim().is_empty() {
        return Err("GitHub OAuth client ID is required".to_string());
    }
    if req.device_code.is_empty() {
        return Err("device_code is required".to_string());
    }

    match crate::ai::auth::poll_device_flow(&req.client_id, &req.device_code)
        .await
        .map_err(|e| e.to_string())?
    {
        Some(tokens) => {
            let token_set = crate::ai::auth::OAuthTokenSet {
                access_token: tokens.access_token,
                refresh_token: tokens.refresh_token,
                expires_at: tokens.expires_at,
            };
            crate::ai::auth::store_tokens(&token_set).map_err(|e| e.to_string())?;
            crate::ai::auth::store_client_id(&req.client_id).map_err(|e| e.to_string())?;
            Ok(CopilotPollLoginResponse { authorized: true })
        }
        None => Ok(CopilotPollLoginResponse { authorized: false }),
    }
}

#[derive(Debug, Serialize)]
pub struct CopilotAuthStatus {
    pub authenticated: bool,
}

#[tauri::command]
pub async fn copilot_auth_status() -> Result<CopilotAuthStatus, String> {
    let authenticated = crate::ai::auth::is_authenticated().map_err(|e| e.to_string())?;
    Ok(CopilotAuthStatus { authenticated })
}

#[tauri::command]
pub async fn copilot_logout() -> Result<(), String> {
    crate::ai::auth::delete_tokens().map_err(|e| e.to_string())
}

File: src/shared/stores/ai/copilot.ts
import { invoke } from "@tauri-apps/api/core";
import type { AIActions, AISlice } from "./types";

export const createCopilotSlice: AISlice<
  Pick<
    AIActions,
    | "loadCopilotAuthStatus"
    | "startCopilotDeviceLogin"
    | "pollCopilotDeviceLogin"
    | "logoutCopilot"
    | "setCopilotClientId"
  >
> = (set, get) => ({
  loadCopilotAuthStatus: async () => {
    try {
      const status = await invoke<{ authenticated: boolean }>("copilot_auth_status");
      set({ copilotAuth: { ...get().copilotAuth, authenticated: status.authenticated } });
    } catch {}
  },

  startCopilotDeviceLogin: async (clientId) => {
    const result = await invoke<{
      device_code: string;
      user_code: string;
      verification_uri: string;
      expires_in: number;
      interval: number;
    }>("copilot_start_device_login", { req: { client_id: clientId } });
    set({ copilotAuth: { ...get().copilotAuth, clientId } });
    return result;
  },

  pollCopilotDeviceLogin: async (clientId, deviceCode) => {
    const result = await invoke<{ authorized: boolean }>("copilot_poll_device_login", {
      req: { client_id: clientId, device_code: deviceCode },
    });
    if (result.authorized) {
      set({ copilotAuth: { ...get().copilotAuth, authenticated: true, clientId } });
    }
    return result.authorized;
  },

  logoutCopilot: async () => {
    await invoke("copilot_logout");
    set({ copilotAuth: { authenticated: false, clientId: get().copilotAuth.clientId } });
  },

  setCopilotClientId: (clientId) => {
    set({ copilotAuth: { ...get().copilotAuth, clientId } });
  },
});
